src/paradigma/heart_rate/feature_extraction.py

```diff
@@ -0,0 +1,288 @@
+        n_samples: int
+    ) -> float:
+    """
+    Calculate the spectral entropy from the normalized power spectral density.
+    """
+    psd_norm = psd / np.sum(psd)
+    spectral_entropy = -np.sum(psd_norm * np.log2(psd_norm)) / np.log2(n_samples)
+    return spectral_entropy
+
+def extract_temporal_domain_features(
+        config: PPGconfig, 
+        df_windowed:pd.DataFrame, 
+        l_quality_stats=['mean', 'std']
+    ) -> pd.DataFrame:
+    """
+    Compute temporal domain features for the ppg signal. The features are added to the dataframe. Therefore the original dataframe is modified, and the modified dataframe is returned.
+
+    Parameters
+    ----------
+
+    config: GaitFeatureExtractionConfig
+        The configuration object containing the parameters for the feature extraction
+    
+    df_windowed: pd.DataFrame
+        The dataframe containing the windowed accelerometer signal
+
+    l_gravity_stats: list, optional
+        The statistics to be computed for the gravity component of the accelerometer signal (default: ['mean', 'std'])
+    
+    Returns
+    -------
+    pd.DataFrame
+        The dataframe with the added temporal domain features.
+    """
+    
+    
+    for stat in l_quality_stats:
+        df_windowed[f'{stat}'] = generate_statistics(
+            sensor_col=df_windowed[config.ppg_colname],
+            statistic=stat
+            )
+    
+
+    df_windowed[f'signal_to_noise'] = compute_signal_to_noise_ratio(df_windowed[config.ppg_colname])  # feature 9
+    df_windowed[f'auto_corr'] = compute_auto_correlation(df_windowed[config.ppg_colname], config.sampling_frequency) # feature 10
+
+    return df_windowed
+
+def extract_spectral_domain_features(
+        config: PPGconfig, 
+        df_windowed:pd.DataFrame
+    ) -> pd.DataFrame:
+    """
+    Calculate the spectral features (dominant frequency, relative power, and spectral entropy)
+    for each segment of a PPG signal using a single Welch's method computation. The features are added to the dataframe. 
+    Therefore the original dataframe is modified, and the modified dataframe is returned.
+
+    Parameters
+    ----------
+    config: PPGconfig
+        The configuration object containing the parameters for the feature extraction
+
+    df_windowed: pd.DataFrame
+        The dataframe containing the windowed ppg signal
+
+    Returns
+    -------
+    pd.DataFrame
+        The dataframe with the added spectral domain features.
+    """
+    fs = config.sampling_frequency
+    ppg_segments = df_windowed[config.ppg_colname]
+    win_len = 3 * fs  # 3-second window for Welch's method
+    window = hamming(win_len, sym=True)  # Hamming window for spectral estimation
+    overlap = win_len // 2  # 50% overlap
+
+    l_dominant_frequencies = []
+    l_relative_powers = []
+    l_spectral_entropies = []
+
+    for segment in ppg_segments:
+        # Compute power spectral density (PSD) once using Welch's method
+        freqs, psd = welch(
```

Comment:
Dit kan ook met vectorizatie dmv `numpy`, maar daar kunnen we later een keer naar kijken samen.

Reply:
lijkt mij goed